File: Queue, Stack, Array/Queue_by_linkedlist.c
#include <stdio.h>
#include <stdlib.h> // for malloc

//node structure
typedef struct linkedNode{
    int val;
    struct LinkedNode*next;
} LinkedNode; // 이러면 struct LinkedNode가 아닌 LinkedNode만 적으면 됌.

//queue structure

typedef struct Queue{
    LinkedNode* front;
    LinkedNode* rear;
    int size;
} Queue;

//make node
LinkedNode* createNode(int value){
    LinkedNode* newnode =  (LinkedNode*)malloc(sizeof(LinkedNode)); // dynamically new node
    newnode->val = value;
    newnode->next =NULL;

    return newnode;
}


//initialization the queue
void initializeQueue(Queue* q){ // pointer로 직접 접근 (참고로 call by reference를 지원하지 않음.)
    q->front = NULL;
    q->rear = NULL;
    q->size = 0;
}


int getSize(Queue* q){
    return q->size;

}

int isEmpty(Queue* q){ // boolean을 반환하지만 정수로 변환해서 반환하게 됨. 마찬가지로 쓸대도, 0이 아닌 모든 값은 true로 판정된다.
    return q->size == 0;
}


//get first
int first(Queue* q){
    if (isEmpty(q)){
        printf("empty queue\n");
        return -1;
    }
    return (*q).front->val; // q->front->val과 동치임.
}

void enqueue(Queue*q, int value){
    LinkedNode* newnode = createNode(value);

    if (isEmpty(q)){
        q->front = q->rear = newnode;
    }
    else{
        q->rear->next = newnode;
        q->rear = newnode;
    }
    q->size++;
}

int dequeue(Queue* q){
    if (isEmpty(q)){
        printf("empty queue\n");
        return -1;
    }
    else{
        LinkedNode* node = q->front;
        int value = node->val;

        if (q->size == 1){
            q->front = q->rear= NULL;
        }
        else{
            q->front = q->front->next;
        }

        q->size--;
        free(node); // release node
        return value;

    }
}

void clear(Queue*q){
    while (!isEmpty(q)){
        dequeue(q); // 계속해서 빼내기.
    }
}

void printQueue(Queue*q){
    if(isEmpty(q)){
        printf("empty queue\n");
    }
    else{
        LinkedNode*temp = q->front;
        while(temp){ // untill temp is not NULL
            printf("value : %d\n", temp->val);
            temp = temp->next;
        }
    }
}



void testQueue() {
    Queue q;
    initializeQueue(&q);

    // Test isEmpty and getSize on empty queue
    printf("Initial empty queue test passed: %s\n", isEmpty(&q) ? "True" : "False");
    printf("Initial queue size: %d\n", getSize(&q));

    // Test enqueue
    enqueue(&q, 10);
    enqueue(&q, 20);
    enqueue(&q, 30);
    printQueue(&q);  // Expected: 10 -> 20 -> 30 -> None

    // Test first
    printf("First element: %d\n", first(&q));

    // Test isEmpty and getSize after enqueues
    printf("Queue is empty: %s\n", isEmpty(&q) ? "True" : "False");
    printf("Queue size: %d\n", getSize(&q));

    // Test dequeue
    printf("Dequeue: %d\n", dequeue(&q));
    printf("Dequeue: %d\n", dequeue(&q));
    printQueue(&q);  // Expected: 30 -> None

    // Test dequeue until empty
    printf("Dequeue: %d\n", dequeue(&q));
    printf("Dequeue: %d\n", dequeue(&q));  // Expected empty queue message
    printQueue(&q);  // Expected: empty queue

    // Test clear
    enqueue(&q, 40);
    enqueue(&q, 50);
    clear(&q);
    printQueue(&q);  // Expected: empty queue
    printf("Queue is empty after clear: %s\n", isEmpty(&q) ? "True" : "False");
    printf("Queue size after clear: %d\n", getSize(&q));

    printf("All tests passed.\n");
}

// Main function to run test cases
int main() {
    testQueue();
    return 0;
}

File: Queue, Stack, Array/Queue_by_linkedlist.py
'''
Queue by singly linked list

is_empty
enqueue
dequeue
first : return the element of the queue(do not remove)
getsize 
clear : remove all
print_queue : print status


'''

class LinkedNode():
    def __init__(self, x):
        self.val = x 
        self.next = None # singly lined list


class Queue:
    def __init__(self):
        self.front = None
        self.rear = None
        self.size = 0


    def getSize(self) -> int:
        return self.size
    

    def is_Empty(self)->bool:
        return self.size ==0

    def first(self)->int:
        if(self.is_Empty()):
            print("empty queue")
        return self.front.val

    def enqueue(self, value) ->None:
        
        #뒤로 넣기.
        
        newnode = LinkedNode(value)
        ## empty이었을 경우
        if self.is_Empty():
            self.front = self.rear = newnode

        #empty가 아니었을 경우.
        else:
            self.rear.next = newnode
        
        self.rear = newnode
        self.size +=1


    def dequeue(self)->int:
        

        ## FIFO이기 때문에 앞에서 뺀다.
        if self.is_Empty():
            print("empty queue")
            return -1
        else:

            node = self.front ## 맨 처음거 불러오기

            #case 1 : 맨 마지막 꺼일 경우,
            if self.size ==1:
                self.front = self.rear = None

            #case 2 : 맨 마지막 성분이 아닐 경우
            else:   
                self.front = self.front.next

            self.size-=1 ## 전체 개수 추적
            return node.val
       

    def clear(self)->None:

        self.front = self.rear =None
        self.size = 0

    def print_queue(self)->None:

        if self.is_Empty():
            print("empty queue")
        else:
            temp = self.front
            while(temp):
                print("value : ", temp.val)
                temp = temp.next
            

# Test cases
def test_queue():
    q = Queue()

    # Test is_Empty and getSize on empty queue
    assert q.is_Empty() == True
    assert q.getSize() == 0
    print("Initial empty queue test passed.")

    # Test enqueue
    q.enqueue(10)
    q.enqueue(20)
    q.enqueue(30)
    q.print_queue()  # Expected: 10 -> 20 -> 30 -> None

    # Test first
    assert q.first() == 10
    print("First element test passed.")

    # Test is_Empty and getSize after enqueues
    assert q.is_Empty() == False
    assert q.getSize() == 3
    print("Queue size and empty check after enqueues passed.")

    # Test dequeue
    assert q.dequeue() == 10
    assert q.dequeue() == 20
    q.print_queue()  # Expected: 30 -> None

    # Test dequeue until empty
    assert q.dequeue() == 30
    assert q.dequeue() == -1  # Expected empty queue message
    q.print_queue()  # Expected: empty queue

    # Test clear
    q.enqueue(40)
    q.enqueue(50)
    q.clear()
    q.print_queue()  # Expected: empty queue
    assert q.is_Empty() == True
    assert q.getSize() == 0
    print("Queue clear test passed.")

    print("All tests passed.")

# Run test cases
if __name__ =="__main__":
    test_queue()


